### PR TITLE
Polyhedron demo: Multiple Viewers

### LINKED ISSUE #2039: Polyhedron_demo : Make several views
## Issue Details
Add an option to have several viewers on the same scene.

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes
This PR (based on #3225) makes it possible to have simultaneously several viewers, allowing to have different viewes on the same scene.

fix #2039

### CODE REVIEW

**Comment by maxGimeno** [comment]
This is a first draft and many optimisations will be performed.

**Comment by afabri** [comment]
I get an error message:

> 
c:\cgal\git\polyhedron\demo\polyhedron\vc2017-multipleviewers\scene_transform_item_autogen\CSF3EUHWGI/moc_Scene_facegraph_transform_item.cpp(99): error C2491: 'Scene_facegraph_transform_item::staticMetaObject': definition of dllimport static data member not allowed

**Comment by maxGimeno** [comment]
It should work now.

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Scene_facegraph_item_k_ring_selection_config.h [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #ifndef SCENE_FACEGRAPH_ITEM_K_RING_SELECTION_CONFIG_H
 #define SCENE_FACEGRAPH_ITEM_K_RING_SELECTION_CONFIG_H
 
-#if defined(scene_poly_item_k_ring_selection_EXPORTS) || (scene_sm_item_k_ring_selection_EXPORTS)
+#if defined(scene_k_ring_selection_EXPORTS)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Scene_facegraph_item_k_ring_selection_config.h]: I thought that kind of error was detected, on Linux, by `-fvisibility=hidden` and `-z defs`. That does not seem to work.

**Comment by maxGimeno** [comment]
I can't find more dll errors on my machines.

**Comment by lrineau** [comment]
Please solve the conflicts (with a merge), so that Travis can test the PR.

**Comment by lrineau** [comment]
Travis is kind of angry with that PR:
https://travis-ci.org/CGAL/cgal/builds/437100056

**Comment by lrineau** [comment]
That did not work. Travis still report issues.

**Comment by afabri** [comment]
With "Look at" the popup window appears twice.

**Comment by afabri** [comment]
How can I set the focus? That is I want to "Look at" a coordinate in the first view, and look at another one in the second view.

**Comment by afabri** [comment]
"Point Inside Polyhedron" no longer selects points.

**Comment by maxGimeno** [comment]
> With "Look at" the popup window appears twice.

Look at is in the View menu for the main viewer, and in every menu of the subviewers. It is not about focus. The camera_positions_plugin however does apply to the currently selected viewer. You can click on or in a viewer to set the focus on it.

**Comment by lrineau** [comment]
This PR has a lot of conflicts with `master`.

**Comment by maxGimeno** [comment]
-[ ] Crash on close.

**Comment by maxGimeno** [comment]
> * Toggling _Draw two sides_ only affects the main viewer. Either it should affect all, or it should affect the currently active viewer (I think the first solution would be less messy). Same problem for _Orthographic projection_.

There is a menu for each sub viewer (the cgal icon) that has the equivalent in each.

**Comment by maxGimeno** [comment]
> * I don't know where this comes from as it doesn't seem related to this PR or to the one it's based onto, but shaded points display of point sets that have both normals and colors is broken: the point set appears to be completely red no matter what the color of the item is and what the RGB colors of the points are. (To reproduce easily, load `b9_training.ply` from the classification examples data and run _Normal estimation_ on it.)

I fixed the problem, but the normal estimation selects points, that's why they are still red, for the most part.

**Comment by sgiraudot** [comment]
> > * Toggling _Draw two sides_ only affects the main viewer. Either it should affect all, or it should affect the currently active viewer (I think the first solution would be less messy). Same problem for _Orthographic projection_.
> 
> There is a menu for each sub viewer (the cgal icon) that has the equivalent in each.
> 
> EDIT: I changed the icon for a more standard menu icon. I hope it will make it obvious enough. Else I don't really know what to do and it will require a discussion with 2 ou 3 people.

It's definitely better with the new icon (indeed with the previous one, I did not expect a useful menu to be there). I think one thing that could help would be to make it clear that some options in "View" only concern the main viewer. But then again, it's not true for all options. Maybe you should split the options of the "View" menu between two menus:

* A _View_ menu that would keep the general options (+ the _Add viewers_ and everything related)
* A _Main Viewer_ menu that would only have the options that apply to the main viewer (and this menu should be exactly the same as the ones of the subviewers). This way, you could also remove the general options (like _Recenter_, etc.) from the subviewer menus (and only keep the options specific to the active subviewer in these dedicated subviewer menus).

> > * I don't know where this comes from as it doesn't seem related to this PR or to the one it's based onto, but shaded points display of point sets that have both normals and colors is broken: the point set appears to be completely red no matter what the color of the item is and what the RGB colors of the points are. (To reproduce easily, load `b9_training.ply` from the classification examples data and run _Normal estimation_ on it.)
> 
> I fixed the problem, but the normal estimation selects points, that's why they are still red, for the most part.

Yes, but the red selection is an expected behaviour, so it's fine now.

All the issues I encountered seem to be fixed, I tested it and it works well. Nice feature!

**Comment by maxGimeno** [comment]
I changed the system of View Menu. Each viewer, including the main viewer, has a system menu with the actions concerning itself, and they are not in the view menu. But When a viewer is maximized, then its system menu is moved in the View menu.

**Comment by sgiraudot** [comment]
I don't see any commit, did you forget to push it or is it a work in progress?

**Comment by maxGimeno** [comment]
Forgot to push indeed.

**Comment by sgiraudot** [comment]
Oh okay, I did not realize that it was for the fullscreen viewer (not necessary the main one), nevermind then (you can leave it that way).

**Comment by maxGimeno** [comment]
It should be fixed.

**Comment by afabri** [comment]
I checked it out and built it with VC2017 in a new build directory. It segfaults when I click on _Add Viewer_

**Comment by maxGimeno** [comment]
Broken again on Windows after rebase.

**Comment by maxGimeno** [comment]
GL Errors in c3t3_item when moving intersection with several viewers.

**Comment by lrineau** [comment]
Doxygen warnings:
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.14-Ic-58/logs_master/Three.log
```
/home/cgal-testsuite/cgal_doc_build/CGAL-4.14-Ic-58/include/CGAL/Three/Viewer_interface.h:114: warning: argument 'sharedWidget' of command @param is not found in the argument list of CGAL::Three::Viewer_interface::Viewer_interface(QWidget *parent, QOpenGLWidget *shared_widget)
/home/cgal-testsuite/cgal_doc_build/CGAL-4.14-Ic-58/include/CGAL/Three/Viewer_interface.h:119: warning: The following parameters of CGAL::Three::Viewer_interface::Viewer_interface(QWidget *parent, QOpenGLWidget *shared_widget) are not documented:
  parameter 'shared_widget'
```

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 2]
```diff
@@ -2546,10 +2550,10 @@ degree(Face_index f) const
     return count;
 }
 
-template <typename P>
+template <typename P> template< typename F>
 void
 Surface_mesh<P>::
-collect_garbage()
+collect_garbage(F& funct)
```
Comment (maxGimeno) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: @sloriot Could you you review this, please ? I'm not very comfortable with adding code in there.

Comment (sloriot) [reply Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: except from the names that I don't like and the typo, it is OK.

Comment (lrineau) [reply Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: And now, @sloriot?

Comment (lrineau) [reply Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: @maxGimeno Can you please point to the new lines of code?

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 3]
```diff
@@ -1282,8 +1282,8 @@ class Surface_mesh
     void collect_garbage();
     
     //undocumented convenience fucntion that allows to get old-index->new-index information
-    template <typename F>
-    void collect_garbage(F& f);
+    template <typename Visitor>
```
Comment (maxGimeno) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: @lrineau @sloriot here

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h [hunk 4]
```diff
@@ -2688,7 +2688,7 @@ collect_garbage(F& funct)
         set_halfedge(f, hmap[halfedge(f)]);
     }
 
-    //apply functor befor invalidating the maps
+    //apply functor before invalidating the maps
```
Comment (maxGimeno) [inline Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: @lrineau @sloriot and here

**Comment by maxGimeno** [comment]
- [ ] Crash on exit when  camera_position_plugin and affine_transformation_plugin are both loaded. (load a polyhedron, add a viewer and quit app.)

File: Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h (hunk 2 above)
Comment (sloriot) [reply Surface_mesh/include/CGAL/Surface_mesh/Surface_mesh.h]: Looks good

**Comment by maxGimeno** [comment]
Conflicts with #3781

**Comment by lrineau** [comment]
Doxygen error:
```
.../CGAL-5.0-Ic-2/include/CGAL/Three/Viewer_interface.h:115: warning: argument 'sharedWidget' of command @param is not found in the argument list of CGAL::Three::Viewer_interface::Viewer_interface(QWidget *parent, QOpenGLWidget *shared_widget)
.../CGAL-5.0-Ic-2/include/CGAL/Three/Viewer_interface.h:120: warning: The following parameters of CGAL::Three::Viewer_interface::Viewer_interface(QWidget *parent, QOpenGLWidget *shared_widget) are not documented:
  parameter 'shared_widget'
```
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-5.0-Ic-2/logs_master/Three.log

**Comment by lrineau** [comment]
Conflicts

**Comment by maxGimeno** [comment]
Successfully tested in  https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-16.shtml

@sloriot if this one could go in master as soon as possible, it would make it far easier to merge branches that impact the demo.

File: Polyhedron/demo/Polyhedron/CMakeLists.txt [hunk 5]
```diff
@@ -397,10 +395,6 @@ add_to_cached_list( CGAL_EXECUTABLE_TARGETS CGAL_Classification )
     NAMESPACE Polyhedron_
     APPEND FILE polyhedron_demo_targets.cmake)
 
-  export(
-    TARGETS
-    NAMESPACE Polyhedron_
-    APPEND FILE polyhedron_demo_targets.cmake)
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CMakeLists.txt]: This commit should have been applied to the release branches as well!

Comment (lrineau) [reply Polyhedron/demo/Polyhedron/CMakeLists.txt]: done

**Pull request merged**